File: data structure and algorithms/Greedy/555.split-concatenated-strings.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <random>
#include <regex>
#include <set>
#include <sstream>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

class Solution {
public:
    string splitLoopedString(vector<string>& strs) {
        vector<string> parts;
        for (string s : strs) {
            string r = s;
            reverse(r.begin(), r.end());
            parts.push_back(max(s, r));
        }
        string best;
        int n = strs.size();
        for (int i = 0; i < n; ++i) {
            string middle;
            for (int j = i + 1; j < n; ++j) middle += parts[j];
            for (int j = 0; j < i; ++j) middle += parts[j];
            vector<string> options = {strs[i], string(strs[i].rbegin(), strs[i].rend())};
            for (const string& cur : options) {
                for (int cut = 0; cut < (int)cur.size(); ++cut) {
                    string cand = cur.substr(cut) + middle + cur.substr(0, cut);
                    best = max(best, cand);
                }
            }
        }
        return best;
    }
};

/*
Interview explanation:
All strings not crossing the final split can independently take their lexicographically larger orientation. The boundary string must be tried in both orientations and all cut positions.

C++ data structures: vector<string> stores chosen orientations; string substrings construct candidates.

Edge cases: using only the boundary string's larger orientation can miss the optimum, so both are tested.

Complexity: O(L^2) candidate construction in the worst case, accepted for constraints; O(L) extra space per candidate.
*/

File: Alibaba/555.split-concatenated-strings.java
import java.util.*;

/**
 * Algorithm:
 * For every string, choose the lexicographically larger orientation for all
 * non-starting pieces. Then try each string as the split piece in both
 * orientations and every cut position, keeping the largest rotation.
 *
 * Java data structures:
 * StringBuilder is used for reversal and candidate construction.
 *
 * Complexity:
 * If total length is L, trying every cut builds O(L) candidates of length L,
 * so time O(L^2) and space O(L).
 */
class Solution {
    public String splitLoopedString(String[] strs) {
        int n = strs.length;
        String[] parts = new String[n];
        for (int i = 0; i < n; i++) {
            String rev = reverse(strs[i]);
            parts[i] = strs[i].compareTo(rev) >= 0 ? strs[i] : rev;
        }

        String best = "";
        for (int i = 0; i < n; i++) {
            StringBuilder middleBuilder = new StringBuilder();
            for (int j = i + 1; j < n; j++) {
                middleBuilder.append(parts[j]);
            }
            for (int j = 0; j < i; j++) {
                middleBuilder.append(parts[j]);
            }
            String middle = middleBuilder.toString();

            String original = strs[i];
            String reversed = reverse(original);
            String[] sources = {original, reversed};
            for (String source : sources) {
                for (int cut = 0; cut < source.length(); cut++) {
                    String candidate = source.substring(cut) + middle + source.substring(0, cut);
                    if (candidate.compareTo(best) > 0) {
                        best = candidate;
                    }
                }
            }
        }
        return best;
    }

    private String reverse(String s) {
        return new StringBuilder(s).reverse().toString();
    }
}

